cookbooks/CPO_induced_anisotropic_viscosity/plugin/cpo_induced_anisotropic_viscosity.h

```diff
@@ -0,0 +1,101 @@
+/*
+  Copyright (C) 2019 - 2020 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_material_model_CPO_AV_3D_h
+#define _aspect_material_model_CPO_AV_3D_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <aspect/material_model/simple.h>
+#include <aspect/material_model/equation_of_state/interface.h>
+#include <aspect/simulator/assemblers/interface.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    template <int dim>
+    class CPO_AV_3D : public MaterialModel::Simple<dim>
+    {
+      public:
+        void initialize() override;
+
+        void evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                       MaterialModel::MaterialModelOutputs<dim> &out) const override;
+
+        bool is_compressible () const override;
+
+        static void declare_parameters (ParameterHandler &prm);
+
+        void parse_parameters (ParameterHandler &prm) override;
+
+        void create_additional_named_outputs(MaterialModel::MaterialModelOutputs<dim> &out) const override;
+
+        void pseudoinverse(LAPACKFullMatrix<double> &A,
+                           LAPACKFullMatrix<double> &A_pinv) const;
```

Comment:
this function can move to the `private` section of this file, and also needs some documentation. What is the input? What is the output? Which algorithm did you use to implement it?